---
title: "Get Identified proteins, Coverage, Length, UpsetR, Venn"
author: "xyz"
date: "2020/4/26"
output: html_document
---

### Filter by q-value

```{r}
library(openxlsx)
df<-read.xlsx("PanamaNCBI2ProteinsFix.xlsx")
df2<-readRDS("panamameMeta2ProteinsFix.xlsx.rds")
suppressMessages(library(dplyr))
df<-filter(df,Contaminant==F,`Exp..q-value:.Combined`<=0.01)
df2<-filter(df2,Contaminant==F,`Exp..q-value:.Combined`<=0.01)
# Using the NCBI soil microbial protein database, 20,779 proteins were detected in 4,344 groups
paste0("Using the NCBI soil microbial protein database ",nrow(df)," proteins were detected ","in ",sum(df$Master=="Master Protein")," groups")
# "170565 proteins were detected using the metagenomic protein database, distributed among 19,298 groups"
paste0("using the metagenomic protein database",nrow(df2),"proteins were detected, ","distributed among ",sum(df2$Master=="Master Protein")," groups")
df<-filter(df,Master=="Master Protein")
df2<-filter(df2,Master=="Master Protein")
# saveRDS(df,"ncbiMasterProtein.rds")
# saveRDS(df2,"metaMasterProtein.rds")
```

### Integrating the eggnog annotation

```{r}
df<-readRDS("ncbiMasterProtein.rds")
df2<-readRDS("metaMasterProtein.rds")

ncbi<-read.table("ncbi.query_seqs.fa.emapper.annotations",
                        quote="",sep="\t",fill=T,header=F)
colnames(ncbi)<-c("query_name","seed eggNOG ortholog","seed ortholog evalue","seed ortholog score",
                 "Predicted taxonomic group","Predicted protein name","Gene Ontology terms",
                 "EC number","KEGG_ko","KEGG_Pathway","KEGG_Module","KEGG_Reaction","KEGG_rclass",
                 "BRITE","KEGG_TC","CAZy","BiGG Reaction","tax_scope","eggNOG OGs","bestOG","COG Functional Category",
                 "eggNOG free text description")
# write.table(ncbi$`Predicted taxonomic group`,"ncbiEggLineage.txt",quote=F,col.names =F,row.names=F)
meta<-read.table("meta.query_seqs.fa.emapper.annotations",
                        quote="",sep="\t",fill=T,header=F)
colnames(meta)<-c("query_name","seed eggNOG ortholog","seed ortholog evalue","seed ortholog score",
                 "Predicted taxonomic group","Predicted protein name","Gene Ontology terms",
                 "EC number","KEGG_ko","KEGG_Pathway","KEGG_Module","KEGG_Reaction","KEGG_rclass",
                 "BRITE","KEGG_TC","CAZy","BiGG Reaction","tax_scope","eggNOG OGs","bestOG","COG Functional Category",
                 "eggNOG free text description")
```

### The Statistics of identified proteins

```{r}
ncbiFound<-df[,72:81]
ncbiFound<-ncbiFound[,]!="Not Found"
# The master proteins identified among all samples by NCBI were 4320
sum(rowSums(ncbiFound[,1:8])>0)
sum(rowSums(ncbiFound[,9:10])>0)

metaFound<-df2[,165:174]
metaFound<-metaFound[,]!="Not Found"
# The master proteins identified among all samples were 18947
sum(sum(rowSums(metaFound[,1:8])>0))
sum(rowSums(metaFound[,9:10])>0)

# Plot
tempDf<-read.table(text = "
Database,p,Count
Public,Protein groups,4320
Public,Proteins,20779
Meta,Protein groups,18947
Meta,Proteins,170565",
header = T,sep = ",",stringsAsFactors=F)
# xlsx::write.xlsx(tempDf,"The amount of identified proteins.xlsx")
library(ggplot2)
ggplot(tempDf,aes(x=p, y=Count, fill=Database)) + 
  geom_bar(stat = "identity",position="dodge",width=0.8, col='black')+
  # Add numbers
  geom_text(aes(label=Count), position=position_dodge(width=0.8), vjust=-0.25, size = 6)+
  ylab("Number of identified \n protein groups/proteins")+
  theme(axis.title.x=element_blank(),text = element_text(size = 30),
        axis.text.x=element_text(colour="black"))+
  ggsave(paste0("The amount of identified proteins",".png"),width = 10.24, height = 7.68,dpi=100)
```

### compare coverage and length

```{r}
library(ggplot2)
tempDf<-data.frame(Coverage=c(df$`Coverage.[%]`,df2$`Coverage.[%]`),
                   Database=rep(c("Public","Meta"),times=c(nrow(df),nrow(df2))))
qplot(Coverage, data = tempDf,geom = "freqpoly", ylab = "Number of identified proteins",binwidth =1,color =Database)+
  theme(text = element_text(size = 30),axis.text = element_text(colour = "black"))+
  ggsave(paste0("compare coverage",".png"),width = 10.24, height = 7.68,dpi=100)
# W = 25026280, p-value < 2.2e-16
wilcox.test(df$`Coverage.[%]`,df2$`Coverage.[%]`)

freq<-data.frame(Coverage=c(df$`Coverage.[%]`,df2$`Coverage.[%]`),
                   Database=rep(c("public","meta"),times=c(nrow(df),nrow(df2))))
freq$value.cut <- cut(freq$Coverage, breaks=c(0, 5, 10,15,20,25,30,100))
freq<-with(freq, table(value.cut, Database, useNA='ifany'))
freq<-as.data.frame(freq)
freq<-tidyr::spread(freq,Database,-value.cut)
# delete NA
freq<-freq[-8,]
colnames(freq)[1]<-"Coverage Range of Proteins %"
# freq<-rbind(freq,cbind(value.cut=freq$value.cut,prop.table(freq[,-1])))
# xlsx::write.xlsx(freq,"compare coverage.xlsx")

library(stringr)
tempDf<-data.frame(Length=c(str_length(df$Sequence),str_length(df2$Sequence)),
                   Database=rep(c("Public","Meta"),times=c(nrow(df),nrow(df2))))
qplot(Length, data = tempDf,geom = "freqpoly", ylab = "Number of identified proteins",binwidth =1,color =Database)+
  theme(text = element_text(size = 30),axis.text = element_text(colour = "black"))+
  ggsave(paste0("compare length of identified proteins",".png"),width = 10.24, height = 7.68,dpi=100)
# Total number of proteins
# public 4320; meta 18947
# Total length of proteins
# public 1693559; meta 7068297
sum(str_length(df$Sequence));sum(str_length(df2$Sequence))
# Average number of amino acids in a protein
# public 389.8616; meta 366.271
mean(str_length(df$Sequence));mean(str_length(df2$Sequence))
# p-value < 2.2e-16
wilcox.test(str_length(df$Sequence),str_length(df2$Sequence))
freq<-data.frame(Length=c(str_length(df$Sequence),str_length(df2$Sequence)),
                   Database=rep(c("public","meta"),times=c(nrow(df),nrow(df2))))
freq$value.cut <- cut(freq$Length, breaks=c(0, 25, 50, 100, 200,400,800,1600,3200,12200))
freq<-with(freq, table(value.cut, Database))
freq<-as.data.frame(freq)
freq<-tidyr::spread(freq,Database,-value.cut)
freq<-cbind(freq,prop.table(as.matrix(freq[,-1]),2))
colnames(freq)[1]<-"Length Range of Proteins %"
xlsx::write.xlsx(freq,"compare length range of proteins.xlsx")
```

### upset plot

```{r}
library(UpSetR)
plotUpset<-function(tempDf,textScale,set_size.angles,fileName){
  drawDf<-as.data.frame(matrix(as.numeric(tempDf),ncol = ncol(tempDf)))
  name<-as.list(colnames(tempDf))
  colnames(drawDf)<-colnames(tempDf)
  png(file = paste0(fileName,".png"),
      width = 1280,height = 720,units = "px")
  print(upset(drawDf,order.by = c("degree","freq"), decreasing = c(TRUE,TRUE),
        # Add a highlight of the intersection between biological and technical repetitions.
        queries = list(list(query = intersects, params = name,color="red", active=TRUE),
                       list(query = intersects, params = name[1:2],color="green", active=TRUE),
                       list(query = intersects, params = name[3:4],color="blue", active=TRUE)
                       ),
        text.scale=textScale,
        # same y scale
        mainbar.y.max=9000,
        # bug, no effect
        set_size.angles=set_size.angles
        ))
  dev.off()
}

# library(ggplot2)
# # https://krassowski.github.io/complex-upset/articles/Examples_R.html
# library(ComplexUpset)
# plotUpset<-function(tempDf,textScale,textAngles){
#   tempDf<-as.data.frame(tempDf[rowSums(tempDf)>0,])
#   name<-colnames(tempDf)
#   upset(
#     tempDf, name, width_ratio=0.1, min_size=10, 
#     # delete set sizes on the left 
#     # set_sizes=F,
#     themes=upset_modify_themes(list('overall_sizes'=theme(axis.text.x=element_text(angle=90)))),
#     base_annotations=list('Intersection size'=intersection_size(text=list(size = textScale,angle=textAngles))),
#     queries=list(upset_query(intersect=name,color='red',fill='red')
#                  # only support highlighting one combination 
#                  # ,upset_query(intersect=name[1:2],color='green',fill='green')
#         ))
# }
# plotUpset(tempDf,5,45)

# public p-rich 4 repetitions in the same scale
tempDf<-ncbiFound[,1:4]
colnames(tempDf)<-paste0("Repeat",1:4)
plotUpset(tempDf,3,90,"public p-rich 4 repetitions in the same scale")

# public p-deficiency 4 repetitions in the same scale
tempDf<-ncbiFound[,5:8]
colnames(tempDf)<-paste0("Repeat",1:4)
plotUpset(tempDf,3,90,"public p-deficiency 4 repetitions in the same scale")

# meta p-rich 4 repetitions in the same scale
tempDf<-metaFound[,1:4]
colnames(tempDf)<-paste0("Repeat",1:4)
plotUpset(tempDf,3,90,"meta p-rich 4 repetitions in the same scale")

# meta p-deficiency 4 repetitions in the same scale
tempDf<-metaFound[,5:8]
colnames(tempDf)<-paste0("Repeat",1:4)
plotUpset(tempDf,3,90,"meta p-deficiency 4 repetitions in the same scale")

# Show the protein detected by Public DB
colSums(ncbiFound)

# Show the protein detected by Meta DB
colSums(metaFound)
```

### Venn Plot

```{r}
null<-suppressMessages(lapply(c("ggplot2","ggforce","limma","venneuler"), require, character.only = TRUE))
plotVenn<-function(tempDf,fileName) {
  # differences between different treatments
  VennDf<-vennCounts(tempDf)
  # convert T and F to 0,1
  drawDf<-matrix(as.numeric(tempDf),ncol = ncol(tempDf))
  # Use Venneuler to generate the data needed for drawing
  drawDf <- venneuler(drawDf)
  drawDf <- data.frame(drawDf$centers, diameters = drawDf$diameters, labels = colnames(tempDf), stringsAsFactors = FALSE)
  ggplot(drawDf) + 
    geom_circle(aes(x0 = x, y0 = y, r = diameters/2, fill = labels), alpha = 0.7) + 
    coord_fixed() +
    theme_void() +
    theme(legend.position = 'bottom',text = element_text(size = 20))+
    labs(fill = NULL) +
    ggsave(fileName,width = 10.24, height = 7.68,dpi=100)
  return(VennDf)
}
tempDf<-ncbiFound[,9:10]
colnames(tempDf)<-c("high","low")
sink("NCBIhighPvsLowP.txt")
plotVenn(tempDf,"NCBIhighPvsLowP.png")
sink()

tempDf<-metaFound[,9:10]
colnames(tempDf)<-c("high","low")
sink("metaHighPvsLowP.txt")
plotVenn(tempDf,"metaHighPvsLowP.png")
sink()
```

